R/is_numeric_linter.R

```diff
@@ -0,0 +1,104 @@
+#' Redirect `is.numeric(x) || is.integer(x)` to just use `is.numeric(x)`
+#'
+#' [is.numeric()] returns `TRUE` when `typeof(x)` is `double` or `integer` --
+#'   testing `is.numeric(x) || is.integer(x)` is thus redundant.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "is.numeric(y) || is.integer(y)",
+#'   linters = is_numeric_linter()
+#' )
+#'
+#' lint(
+#'   text = "class(z) %in% c('numeric', 'integer')",
+#'   linters = is_numeric_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "is.numeric(y) || is.factor(y)",
+#'   linters = is_numeric_linter()
+#' )
+#'
+#' lint(
+#'   text = "class(z) %in% c('numeric', 'integer', 'factor')",
+#'   linters = is_numeric_linter()
+#' )
+#'
+#' @evalRd rd_tags("is_numeric_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+is_numeric_linter <- function() {
+  # TODO(michaelchirico): this should also cover is.double(x) || is.integer(x)
+  # TODO(#1636): is.numeric(x) || is.integer(x) || is.factor(x) is also redundant
+  is_numeric_expr <- "expr[1][SYMBOL_FUNCTION_CALL[text() = 'is.numeric']]"
+  is_integer_expr <- "expr[1][SYMBOL_FUNCTION_CALL[text() = 'is.integer']]"
+
```

Comment:
Should there be a comment explaining that we don't need to cover `any(class(x) == c("numeric", "integer"))` in this linter?

``` r
library(lintr)

lint(
  text = 'any(class(x) == c("numeric", "integer"))',
  linters = linters_with_tags(tags = NULL)
)
#> <text>:1:5: warning: [class_equals_linter] Instead of comparing class(x) with ==, use inherits(x, 'class-name') or is.<class> or is(x, 'class')
#> any(class(x) == c("numeric", "integer"))
#>     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

<sup>Created on 2022-10-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
we might want to lint this here directly. I'll make a note in the docs for now.